File: VisionTesting.cpp
#include <nivision.h>
//#include <niimaq.h>
#include <stdlib.h>
#include <math.h>
#include <iostream>
#include <fstream>
#include <string>
#include "WPILib.h"
#include "CameraControl.h"
#include "PotentiometerMonitor.cpp"
#include "TurretOutput.cpp"

using namespace std;

/* Port configuration for sensors and actuators. */
	#define MANIPULATOR_JOYSTICK_PORT 1

	#define LEFT_DRIVE_MOTOR_PORT 8
	#define RIGHT_DRIVE_MOTOR_PORT 4
	
	#define SHOOTER_MOTOR_PORT 3
	#define TURRET_MOTOR_PORT 2
	#define PICKUP_MOTOR_PORT 6
	#define FEED_MOTOR_PORT 7

	#define POTENTIOMETER_PORT 7
	#define GYRO_PORT 1

	#define LIGHT_RELAY_PORT 1

/* Button configuration. */
	

/* Sensor configuration. */
	#define TURRET_POWER -0.4

class VisionTesting : public SimpleRobot {
	Joystick joystickManipulator;
	Victor motorTurret;
	
	AnalogChannel analogPotentiometer;
	Gyro gyro;
	
	Task cameraTask;
	
	PotentiometerMonitor *potentiometerMonitor;
	TurretOutput *turretOutput;
	PIDController *turretController;
	
	Relay relayLight;
	
	public:
		VisionTesting (void):
			joystickManipulator(MANIPULATOR_JOYSTICK_PORT),
			motorTurret(TURRET_MOTOR_PORT),
			analogPotentiometer(POTENTIOMETER_PORT),
			gyro(GYRO_PORT),
			cameraTask("cameraTask", (FUNCPTR)(CameraProcess)),
			relayLight(LIGHT_RELAY_PORT, Relay::kForwardOnly)
		{
			GetWatchdog().SetEnabled(false);
			
			potentiometerMonitor = new PotentiometerMonitor(&analogPotentiometer);
			turretOutput = new TurretOutput(&motorTurret);
			turretController = new PIDController(0.009, 0, 0.0011, potentiometerMonitor, turretOutput);
			turretController->SetSetpoint(90);
		}
				
		void Autonomous (void) {
			GetWatchdog().SetEnabled(false);
		}

		void OperatorControl (void) {
			GetWatchdog().SetEnabled(true);
			GetWatchdog().Feed();
			
			double gyroPosition;
			double currentPosition;
			double lastPosition;
			
			CameraData &cameraData = CameraData::GetInstance();
			
			Timer *runTimer = new Timer();
			runTimer->Start();
			
			Timer *gyroTimer = new Timer();
			gyroTimer->Start();
			
			float turretPower = 0;
			
			bool wasEnabled = false;
			
			ofstream logFile;
			char buffer[128];
			
			sprintf(buffer, "robot-%f.log", Timer::GetFPGATimestamp());
			logFile.open(buffer);
			
			CameraData::GetInstance().enabled = true;
			cameraTask.Start();
			
			bool lightOn = false;
			
			string line;
			ifstream targetFile;
			char *a;
			
			bool wasAuto = false;
			float currPoint = 90;
			
			while (IsOperatorControl() && IsEnabled()) {
				GetWatchdog().Feed();
				
				cameraData.log = joystickManipulator.GetRawButton(5);
				
				if(joystickManipulator.GetRawButton(9)) {
					lightOn = !lightOn;
					if(lightOn)
						relayLight.Set(Relay::kOn);
					else
						relayLight.Set(Relay::kOff);
				}
				
				if(joystickManipulator.GetRawButton(8)) {
					targetFile.open("target.txt");
					if(targetFile.is_open()) {
						while(targetFile.good()) {
							getline(targetFile, line);
							a = new char[line.size() + 1];
							a[line.size()] = 0;
							memcpy(a, line.c_str(), line.size());
						}
					}
					currPoint = atof(a);
					turretController->SetSetpoint(currPoint);
					targetFile.close();
				}
				
				if(joystickManipulator.GetRawButton(2)) {
					wasAuto = true;
					turretController->SetSetpoint(potentiometerMonitor->PIDGet() + cameraData.offsetAngleX);
				} else if(wasAuto) {
					wasAuto = false;
					turretController->SetSetpoint(currPoint);
				}
				
				if(joystickManipulator.GetRawButton(4)) {
					gyroPosition = 0;
					lastPosition = 0;
					runTimer->Reset();
					gyroTimer->Reset();
					gyro.Reset();
				}
				
				currentPosition = gyro.GetAngle();
				gyroPosition += currentPosition - lastPosition;
				lastPosition = currentPosition;
				
				//gyroPosition -= gyroTimer->Get() * 0.0238095238;
				gyroTimer->Reset();
				
				SmartDashboard::GetInstance()->PutDouble("gyroPosition", gyroPosition);
				sprintf(buffer, "[%f] %f\n", runTimer->Get(), gyroPosition);
				logFile << buffer;
				
				if(joystickManipulator.GetRawButton(1)) {
					if(!wasEnabled) {
						turretController->Enable();
						wasEnabled = true;
					} else {
						SmartDashboard::GetInstance()->PutDouble("lastOutput", turretOutput->lastOutput);
					}
				} else {
					if(wasEnabled) {
						turretController->Disable();
						wasEnabled = false;
					} else {
						SmartDashboard::GetInstance()->PutDouble("lastOutput", 0);
					}
					
					turretPower = joystickManipulator.GetX();
					if(turretPower < -0.2 || turretPower > 0.2)
						motorTurret.Set(joystickManipulator.GetX() * TURRET_POWER);
					else
						motorTurret.Set(0);
				}
				
				SmartDashboard::GetInstance()->PutDouble("centerX", cameraData.centerX);
				SmartDashboard::GetInstance()->PutDouble("centerY", cameraData.centerY);
				SmartDashboard::GetInstance()->PutDouble("offsetAngleX", cameraData.offsetAngleX);
				SmartDashboard::GetInstance()->PutDouble("offsetAngleY", cameraData.offsetAngleY);
				SmartDashboard::GetInstance()->PutDouble("frames", CameraData::GetInstance().frames);
				SmartDashboard::GetInstance()->PutDouble("fps", CameraData::GetInstance().fps);
				
				SmartDashboard::GetInstance()->PutDouble("analogPotentiometer", analogPotentiometer.GetVoltage());
				SmartDashboard::GetInstance()->PutDouble("PotentiometerToDegrees", potentiometerMonitor->PIDGet());
				SmartDashboard::GetInstance()->PutDouble("setPoint", turretController->GetSetpoint());
			}
			
			relayLight.Set(Relay::kOff);
			
			logFile.close();
			
			gyroTimer->Stop();
			runTimer->Stop();
			
			CameraData::GetInstance().enabled = false;
			cameraTask.Stop();
			
			GetWatchdog().SetEnabled(false);
		}
		
		void RobotInit(void) {
			printf("Hello, ninja h4X0r.");
		}
		
		void Disabled (void) {
			GetWatchdog().SetEnabled(false);
		}
	};

START_ROBOT_CLASS(VisionTesting);

File: src/com/_604robotics/robot2012/visiontesting/VisionTesting.java
package com._604robotics.robot2012.visiontesting;

import com._604robotics.robot2012.Aiming.Aiming;
import com._604robotics.robot2012.Aiming.Point3d;
import com._604robotics.robot2012.visiontesting.camera.CameraInterface;
import com._604robotics.robot2012.visiontesting.camera.RemoteCameraTCP;
import com._604robotics.robot2012.visiontesting.configuration.ButtonConfiguration;
import com._604robotics.robot2012.visiontesting.configuration.CameraConfiguration;
import com._604robotics.robot2012.visiontesting.configuration.PortConfiguration;
import com.sun.squawk.util.MathUtils;
import edu.wpi.first.wpilibj.*;
import edu.wpi.first.wpilibj.camera.AxisCamera;
import edu.wpi.first.wpilibj.image.BinaryImage;
import edu.wpi.first.wpilibj.image.ColorImage;
import edu.wpi.first.wpilibj.smartdashboard.SmartDashboard;

public class VisionTesting extends SimpleRobot {
    public final Joystick controller = new Joystick(PortConfiguration.XBOX_CONTROLLER_PORT);
    public final Victor turret = new Victor(PortConfiguration.TURRET_MOTOR_PORT);
    
    public final AnalogChannel potentiometer = new AnalogChannel(PortConfiguration.POTENTIOMETER_PORT);
    public final Gyro gyro = new Gyro(PortConfiguration.GYRO_PORT);
    
    public final Relay light = new Relay(PortConfiguration.LIGHT_RELAY_PORT, Relay.Direction.kForward);
   
    public final PotentiometerMonitor potentiometerMonitor = new PotentiometerMonitor(potentiometer);
    public final TurretOutput turretOutput = new TurretOutput(turret);
    public final PIDController turretController = new PIDController(0.009, 0, 0.0011, potentiometerMonitor, turretOutput);
    
    public final CameraInterface cameraInterface = new RemoteCameraTCP();
    
    public final Aiming aiming = new Aiming();
    
    public VisionTesting () {
        this.getWatchdog().setEnabled(false);
    }
    
    public void autonomous () {
        light.set(Relay.Value.kOn);
        
        System.out.println("WAITING");
        
        Timer timer = new Timer();
        timer.start();
        
        while(timer.get() < 3);
        timer.stop();
        
        System.out.println("READY");
        System.out.println("CAPTURING");
        
        AxisCamera camera = AxisCamera.getInstance();
        
        while (!camera.freshImage());
        
        try {
            ColorImage img = camera.getImage();
            img.write("/img.png");
            System.out.println("CAPTURED");
            
            BinaryImage binImage = img.thresholdHSL(CameraConfiguration.THRESHOLD_MIN_HUE, CameraConfiguration.THRESHOLD_MAX_HUE, CameraConfiguration.THRESHOLD_MIN_SATURATION, CameraConfiguration.THRESHOLD_MAX_SATURATION, CameraConfiguration.THRESHOLD_MIN_LUMINANCE, CameraConfiguration.THRESHOLD_MAX_LUMINANCE);
            System.out.println(Integer.toString(binImage.getNumberParticles()));
            binImage.write("/bin.png");
            System.out.println("BINARY");
            
            BinaryImage convexHull = binImage.convexHull(true);
            convexHull.write("/convex.png");
            System.out.println("CONVEX");
            
            BinaryImage cherryPicked = convexHull.removeSmallObjects(true, 3);
            cherryPicked.write("/cherry.png");
            System.out.println("CHERRY");
            
            System.out.println(Double.toString(cherryPicked.getParticleAnalysisReport(0).center_mass_x));
            
            img.free();
            binImage.free();
            convexHull.free();
            cherryPicked.free();
            
            System.out.println("FREED");
       } catch (Exception ex) {
            System.err.println("ERROR!");
            System.err.println(ex.getMessage());
            ex.printStackTrace();
        }
        
        light.set(Relay.Value.kOff);
        System.out.println("DONE");
    }
    
    public void operatorControl () {
        Timer runTimer = new Timer();
        runTimer.start();
        
        boolean lightOn = false;
        Timer flickTimer = new Timer();
        flickTimer.start();
        
        Timer logTimer = new Timer();
        logTimer.start();
        
        double turretPower = 0;
        boolean wasAuto = false;
        
	double gyroCurrentPosition = 0;
	double gyroProcessedPosition = 0;
	double gyroLastPosition = 0;
        
        Point3d[] points;
        cameraInterface.begin();
        
        while(this.isOperatorControl() && this.isEnabled()) {
            gyroCurrentPosition = gyro.getAngle();
            gyroProcessedPosition += gyroCurrentPosition - gyroLastPosition;
            gyroLastPosition = gyroCurrentPosition;
            
            if(gyroProcessedPosition >= 360) gyroProcessedPosition -= 360;
            if(gyroProcessedPosition < 0) gyroProcessedPosition += 360;
            
            if (controller.getRawButton(ButtonConfiguration.LIGHT_SWITCH) && flickTimer.get() > 1) {
                lightOn = !lightOn;

                if (lightOn)
                    light.set(Relay.Value.kOn);
                else
                    light.set(Relay.Value.kOff);
                
                System.out.println((lightOn) ? "Light ON" : "Light OFF");
                
                flickTimer.reset();
            }

            if (controller.getRawButton(ButtonConfiguration.AUTOMATED_CONTROL)) {
                wasAuto = true;
                turretController.enable();
            } else {
                if (wasAuto) {
                    wasAuto = false;
                    turretController.disable();
                }

                turretPower = controller.getRawAxis(1); // XBOX: Left stick X-ax

                if (turretPower < -0.2 || turretPower > 0.2)
                    turret.set(turretPower * -0.4);
                else
                    turret.set(0D);
            }
            
            SmartDashboard.putDouble("gyro", gyroProcessedPosition);
            
            SmartDashboard.putDouble("potentiometer", potentiometerMonitor.pidGet());
            SmartDashboard.putDouble("setpoint", turretController.getSetpoint());
            
            points = cameraInterface.getTargets();
            
            for (int i = 0; i < points.length; i++)
                System.out.println("x: " + points[i].x + ", y: " + points[i].y + ", z: " + points[i].z);
            
            System.out.println(" - UPS: " + ((RemoteCameraTCP) cameraInterface).getUPS() + " - ");
            System.out.println("-------------------------------");
            
            if (controller.getRawButton(ButtonConfiguration.LOAD_SETPOINT_FROM_VISION))
                turretController.setSetpoint(Math.toDegrees(MathUtils.asin(points[0].x / points[0].z)) - gyroProcessedPosition);
            
            runTimer.reset();
        }
        
        logTimer.stop();
        flickTimer.stop();
        runTimer.stop();
        
        cameraInterface.end();
    }
    
    public void disabled () {
        
    }
    
    public void robotInit () {
        System.out.println("Hello, ninja h4X0r.");
    }
}